src/worker_proxy.ts

```diff
@@ -0,0 +1,121 @@
+import * as Comlink from 'comlink';
+import type { api } from './lib_worker.ts';
+// @ts-ignore (esbuild-plugin-inline-worker will rewrite this import)
+import DedicatedWorker from './h5wasm.worker.ts';
+
+import { ACCESS_MODES } from './hdf5_hl.ts';
+import type { File as H5WasmFile, Group, Dataset, Datatype, BrokenSoftLink } from './hdf5_hl.ts';
+export type { H5WasmFile, Group, Dataset, Datatype, BrokenSoftLink };
+
+type ACCESS_MODESTRING = keyof typeof ACCESS_MODES;
+
+const worker = new DedicatedWorker(); // new Worker('./worker.js');
+const remote = Comlink.wrap(worker) as Comlink.Remote<typeof api>;
+
+export class GroupProxy {
+  proxy: Comlink.Remote<Group>;
+  file_id: bigint;
+  constructor(proxy: Comlink.Remote<Group>, file_id: bigint) {
+    this.proxy = proxy;
+    this.file_id = file_id;
+  }
+
+  async keys() {
+    return await this.proxy.keys();
+  }
+
+  async paths() {
+    return await this.proxy.paths();
+  } 
+
+  async get(name: string = "/") {
```

Comment:
OK, I think I understand, you have to basically reimplement your own `Group#get` method that returns proxied entity objects because the objects returned by the original method loose all their methods.

The downside I see is that we're basically exposing a `Group` class from the worker that doesn't quite work as designed. Also, when we get an entity at a given path, we end up either with a "native" h5wasm entity (`Dataset`, `Datatype`, etc.) or with a `GroupProxy`, which might be confusing.

I have an idea for a slightly different approach:

- Instead of the "native" h5wasm `Group`, `Dataset`, etc. classes, expose superclasses that forbid calling all unsupported methods, like `Group#get`.
- Provide a functional utility, maybe `get_entity(path)` that implements the same logic as `Group#get` (including calling `Module#get_type`, `Module#get_external_link`, etc.) but returns instances of the proxied superclasses.

```ts
class WorkerGroup extends Group {
  public override get(path: string) {
    throw new Error('Method `get` not supported);
  }
}

Comlink.expose({
  ready: h5wasm.ready,
  WorkerGroup
});
```

```ts
export async function get_entity(fileId: string, entityPath: string): Promise<WorkerGroup | WorkerDataset | ...> {
  const module = await remote.ready;
  const kind = await module.get_type(fileId, entityPath);
  switch (kind) {  ...  }
}
```

Reply:
With this superclass approach, we can also override some methods to return [`Transferable`](https://github.com/googlechromelabs/comlink?tab=readme-ov-file#comlinktransfervalue-transferables-and-comlinkproxyvalue) values to avoid copying buffers and increase performance.

Reply:
I agree that it will be important for performance to refine this so it transfers large objects instead of copying them.

For the Group proxy though, I would hate to lose the ability to use `Group.get`... I feel like that's an important part of the current API, particularly when this is being used in an interactive session, where you might want to explore a Group or its children without having to reconstruct the whole HDF path every time (and append to it).  I understand the reasons for including a `get_entity(<absolute_hdf_path>)` method, but can't we do both?

Reply:
Fair enough, it's true that they are not incompatible approaches. I guess you could just call `get_entity` from `GroupProxy#get` and if the result is a `WorkerGroup`, wrap it in another `GroupProxy`. :+1: